<html><h1>Validation FAILED.</h1><p>Build Time: 30 minutes 53 seconds 229 ms</p><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/com-liferay-commerce/tree/7.1.x">7.1.x</a><br/>Branch GIT ID: <a href="https://github.com/liferay/com-liferay-commerce/commit/33857084450ec106ccb267826ce2acaf16e46be2">33857084450ec106ccb267826ce2acaf16e46be2</a></p><h6>Task Summary:</h6><ul><li>subrepository-compile-jdk8 -  :heavy_check_mark:</li><li>subrepository-semantic-versioning-jdk8 -  :heavy_check_mark:</li><li>subrepository-pmd-jdk8 -  :heavy_check_mark:</li><li>subrepository-source-format-jdk8 -  :x:<pre><code>     [echo] java.lang.Exception: Found 4 formatting issues:
     [echo] 1: ./commerce-product-content-category-web/build.gradle expected:&lt;...om.liferay.commerce.[media.api", version: "1.1.0"
     [echo] 	compileOnly group: "com.liferay.commerce", name: "com.liferay.commerce.product.api", version: "6.0].0"
     [echo] 	compileOnly gro...&gt; but was:&lt;...om.liferay.commerce.[product.api", version: "6.0.0"
     [echo] 	compileOnly group: "com.liferay.commerce", name: "com.liferay.commerce.media.api", version: "1.1].0"
     [echo] 	compileOnly gro...&gt;
     [echo] 2: ./commerce-product-asset-categories-web/build.gradle expected:&lt;...om.liferay.commerce.[media.api", version: "1.1.0"
     [echo] 	compileOnly group: "com.liferay.commerce", name: "com.liferay.commerce.product.api", version: "6.0].0"
     [echo] 	compileOnly gro...&gt; but was:&lt;...om.liferay.commerce.[product.api", version: "6.0.0"
     [echo] 	compileOnly group: "com.liferay.commerce", name: "com.liferay.commerce.media.api", version: "1.1].0"
     [echo] 	compileOnly gro...&gt;
     [echo] 3: ./commerce-product-asset-categories-navigation-web/build.gradle expected:&lt;...om.liferay.commerce.[media.api", version: "1.1.0"
     [echo] 	compileOnly group: "com.liferay.commerce", name: "com.liferay.commerce.product.api", version: "6.0].0"
     [echo] 	compileOnly gro...&gt; but was:&lt;...om.liferay.commerce.[product.api", version: "6.0.0"
     [echo] 	compileOnly group: "com.liferay.commerce", name: "com.liferay.commerce.media.api", version: "1.1].0"
     [echo] 	compileOnly gro...&gt;
     [echo] 4: ./commerce-product-definitions-web/build.gradle expected:&lt;...om.liferay.commerce.[media.api", version: "1.1.0"
     [echo] 	compileOnly group: "com.liferay.commerce", name: "com.liferay.commerce.product.api", version: "6.0].0"
     [echo] 	compileOnly gro...&gt; but was:&lt;...om.liferay.commerce.[product.api", version: "6.0.0"
     [echo] 	compileOnly group: "com.liferay.commerce", name: "com.liferay.commerce.media.api", version: "1.1].0"
     [echo] 	compileOnly gro...&gt;
     [echo] 
     [echo] 	at com.liferay.source.formatter.SourceFormatter.format(SourceFormatter.java:426)
     [echo] 	at com.liferay.source.formatter.SourceFormatter.main(SourceFormatter.java:254)
     [echo] 
     [echo] FAILURE: Build failed with an exception.
     [echo] 
     [echo] * What went wrong:
     [echo] Execution failed for task ':checkSourceFormatting'.</code></pre></li><li>central-requirements-jdk8/subrepository-unit-jdk8 -  :heavy_check_mark:</li></ul><h5>For full console, click <a href="https://test-1-11.liferay.com/job/test-subrepository-acceptance-pullrequest-validation(7.1.x)/121//consoleText">here</a>.</h5><hr/><div><h6>Test Results:</h6><div><div><p>46 Tests Passed.<br/>0 Tests Failed.</p></div></div><h5>For all test results, click <a href="https://test-1-11.liferay.com/job/test-subrepository-acceptance-pullrequest-validation(7.1.x)/121//testReport">here</a>.</h5></div></html>

--- a/commerce-media-impl/src/main/java/com/liferay/commerce/media/internal/DefaultCommerceMediaResolver.java
+++ b/commerce-media-impl/src/main/java/com/liferay/commerce/media/internal/DefaultCommerceMediaResolver.java
@@ -14,6 +14,8 @@
 
 package com.liferay.commerce.media.internal;
 
+import com.liferay.asset.kernel.model.AssetCategory;
+import com.liferay.asset.kernel.service.AssetCategoryLocalService;
 import com.liferay.commerce.account.model.CommerceAccount;
 import com.liferay.commerce.account.util.CommerceAccountHelper;
 import com.liferay.commerce.media.CommerceMediaResolver;
@@ -55,6 +57,7 @@ import com.liferay.portal.kernel.util.Portal;
 import com.liferay.portal.kernel.util.StringUtil;
 import com.liferay.portal.kernel.util.URLCodec;
 import com.liferay.portal.kernel.util.WebKeys;
+import com.liferay.portlet.asset.service.permission.AssetCategoryPermission;
 
 import java.io.IOException;
 
@@ -128,7 +131,7 @@ public class DefaultCommerceMediaResolver implements CommerceMediaResolver {
 			boolean secure)
 		throws PortalException {
 
-		StringBundler sb = new StringBundler(13);
+		StringBundler sb = new StringBundler(10);
 
 		sb.append(_portal.getPathModule());
 		sb.append(StringPool.SLASH);
@@ -165,24 +168,10 @@ public class DefaultCommerceMediaResolver implements CommerceMediaResolver {
 
 		String className = cpAttachmentFileEntry.getClassName();
 
-		if (className.equals(CPDefinition.class.getName())) {
-			sb.append("/products/");
-
-			CPDefinition cpDefinition =
-				_cpDefinitionLocalService.getCPDefinition(
-					cpAttachmentFileEntry.getClassPK());
-
-			sb.append(cpDefinition.getCProductId());
-
-			sb.append(StringPool.SLASH);
-
-			Map<Locale, String> titleMap = cpDefinition.getUrlTitleMap();
-
-			sb.append(titleMap.get(siteDefaultLocale));
-		}
-		else {
-			throw new UnsupportedOperationException();
-		}
+		sb.append(
+			setUrl(
+				className, cpAttachmentFileEntry.getClassPK(),
+				siteDefaultLocale));
 
 		sb.append(StringPool.SLASH);
 		sb.append(cpAttachmentFileEntry.getFileEntryId());
@@ -232,34 +221,22 @@ public class DefaultCommerceMediaResolver implements CommerceMediaResolver {
 			return;
 		}
 
-		long cProductId = GetterUtil.getLong(pathArray[1]);
+		long groupId = getGroupId(
+			pathArray[0], GetterUtil.getLong(pathArray[1]), httpServletRequest);
 
-		CProduct cProduct = _cProductLocalService.fetchCProduct(cProductId);
-
-		if (cProduct == null) {
+		if (groupId == 0) {
 			httpServletResponse.sendError(HttpServletResponse.SC_NOT_FOUND);
 
 			return;
 		}
 
 		try {
-			setCPRules(httpServletRequest, cProduct.getGroupId());
-
-			if (!_cpDefinitionModelResourcePermission.contains(
-					PermissionThreadLocal.getPermissionChecker(),
-					cProduct.getPublishedCPDefinitionId(), ActionKeys.VIEW)) {
-
-				httpServletResponse.sendError(HttpServletResponse.SC_NOT_FOUND);
-
-				return;
-			}
-
 			FileEntry fileEntry = getFileEntry(httpServletRequest);
 
 			if (fileEntry == null) {
 				sendDefaultMediaBytes(
 					httpServletRequest, httpServletResponse, contentDisposition,
-					cProduct.getGroupId());
+					groupId);
 
 				return;
 			}
@@ -309,6 +286,48 @@ public class DefaultCommerceMediaResolver implements CommerceMediaResolver {
 
 			return null;
 		}
+	}
+
+	protected long getGroupId(
+		String mediaType, long primaryKey,
+		HttpServletRequest httpServletRequest) {
+
+		if (mediaType.equals("asset-categories")) {
+			AssetCategory assetCategory =
+				_assetCategoryLocalService.fetchCategory(primaryKey);
+
+			try {
+				if (AssetCategoryPermission.contains(
+						PermissionThreadLocal.getPermissionChecker(),
+						assetCategory, ActionKeys.VIEW)) {
+
+					return assetCategory.getGroupId();
+				}
+			}
+			catch (PortalException pe) {
+				_log.error(pe, pe);
+			}
+		}
+		else if (mediaType.equals("products")) {
+			CProduct cProduct = _cProductLocalService.fetchCProduct(primaryKey);
+
+			try {
+				setCPRules(httpServletRequest, cProduct.getGroupId());
+
+				if (_cpDefinitionModelResourcePermission.contains(
+						PermissionThreadLocal.getPermissionChecker(),
+						cProduct.getPublishedCPDefinitionId(),
+						ActionKeys.VIEW)) {
+
+					return cProduct.getGroupId();
+				}
+			}
+			catch (PortalException pe) {
+				_log.error(pe, pe);
+			}
+		}
+
+		return 0;
 	}
 
 	protected void sendDefaultMediaBytes(
@@ -379,8 +398,51 @@ public class DefaultCommerceMediaResolver implements CommerceMediaResolver {
 		CPRulesThreadLocal.setCPRules(cpRules);
 	}
 
+	protected String setUrl(String className, long classPK, Locale locale)
+		throws PortalException {
+
+		StringBundler sb = new StringBundler(4);
+
+		if (className.equals(CPDefinition.class.getName())) {
+			sb.append("/products/");
+
+			CPDefinition cpDefinition =
+				_cpDefinitionLocalService.getCPDefinition(classPK);
+
+			sb.append(cpDefinition.getCProductId());
+
+			sb.append(StringPool.SLASH);
+
+			Map<Locale, String> titleMap = cpDefinition.getUrlTitleMap();
+
+			sb.append(titleMap.get(locale));
+		}
+		else if (className.equals(AssetCategory.class.getName())) {
+			sb.append("/asset-categories/");
+
+			AssetCategory assetCategory =
+				_assetCategoryLocalService.getAssetCategory(classPK);
+
+			sb.append(assetCategory.getCategoryId());
+
+			sb.append(StringPool.SLASH);
+
+			Map<Locale, String> titleMap = assetCategory.getTitleMap();
+
+			sb.append(titleMap.get(locale));
+		}
+		else {
+			throw new UnsupportedOperationException();
+		}
+
+		return sb.toString();
+	}
+
 	private static final Log _log = LogFactoryUtil.getLog(
 		DefaultCommerceMediaResolver.class);
+
+	@Reference
+	private AssetCategoryLocalService _assetCategoryLocalService;
 
 	@Reference
 	private CommerceAccountHelper _commerceAccountHelper;

--- a/commerce-product-service/src/main/java/com/liferay/commerce/product/service/impl/CPAttachmentFileEntryServiceImpl.java
+++ b/commerce-product-service/src/main/java/com/liferay/commerce/product/service/impl/CPAttachmentFileEntryServiceImpl.java
@@ -14,6 +14,7 @@
 
 package com.liferay.commerce.product.service.impl;
 
+import com.liferay.asset.kernel.model.AssetCategory;
 import com.liferay.commerce.product.constants.CPActionKeys;
 import com.liferay.commerce.product.constants.CPConstants;
 import com.liferay.commerce.product.model.CPAttachmentFileEntry;
@@ -33,6 +34,7 @@ import com.liferay.portal.kernel.service.ServiceContext;
 import com.liferay.portal.kernel.util.OrderByComparator;
 import com.liferay.portal.kernel.util.Portal;
 import com.liferay.portal.spring.extender.service.ServiceReference;
+import com.liferay.portlet.asset.service.permission.AssetCategoryPermission;
 
 import java.util.ArrayList;
 import java.util.List;
@@ -91,10 +93,20 @@ public class CPAttachmentFileEntryServiceImpl
 			long cpDefinitionClassNameId = _portal.getClassNameId(
 				CPDefinition.class);
 
+			long assetCategoryClassNameId = _portal.getClassNameId(
+				AssetCategory.class);
+
 			if (cpDefinitionClassNameId ==
 					cpAttachmentFileEntry.getClassNameId()) {
 
 				_cpDefinitionModelResourcePermission.check(
+					getPermissionChecker(), cpAttachmentFileEntry.getClassPK(),
+					ActionKeys.VIEW);
+			}
+			else if (assetCategoryClassNameId ==
+						cpAttachmentFileEntry.getClassNameId()) {
+
+				AssetCategoryPermission.check(
 					getPermissionChecker(), cpAttachmentFileEntry.getClassPK(),
 					ActionKeys.VIEW);
 			}
@@ -198,10 +210,20 @@ public class CPAttachmentFileEntryServiceImpl
 			long cpDefinitionClassNameId = _portal.getClassNameId(
 				CPDefinition.class);
 
+			long assetCategoryClassNameId = _portal.getClassNameId(
+				AssetCategory.class);
+
 			if (cpDefinitionClassNameId ==
 					cpAttachmentFileEntry.getClassNameId()) {
 
 				_cpDefinitionModelResourcePermission.check(
+					getPermissionChecker(), cpAttachmentFileEntry.getClassPK(),
+					ActionKeys.VIEW);
+			}
+			else if (assetCategoryClassNameId ==
+						cpAttachmentFileEntry.getClassNameId()) {
+
+				AssetCategoryPermission.check(
 					getPermissionChecker(), cpAttachmentFileEntry.getClassPK(),
 					ActionKeys.VIEW);
 			}

